### PR TITLE
Add length for instructions and expressions

### PR DESCRIPTION
**Pull request by i1i1**
Add check for number of instructions and expressions in transaction payload. If it exceeds constant (currently 4096) then transaction won't be accepted.

Remove double boxing for some instruction variants.

Signed-off-by: i1i1 <vanyarybin1@live.ru>

### Benefits

It will prevent peers from overflow of instructions from bad peers.

### Possible Drawbacks

It may take some time to count all instructions and expression as it involves full graph traversal and going through a lot of references which are not located near each other.

### CODE REVIEW

File: iroha/src/torii.rs [hunk 1]
```diff
@@ -128,12 +134,19 @@ struct ToriiState {
     sumeragi: Arc<RwLock<Sumeragi>>,
 }
 
+/// Maximum number of bytes in raw transaction. Used to prevent from DOS attacks.
+pub const MAX_TRANSACTION_SIZE: usize = 32 * 1024;
```
Comment (eadventurous) [inline iroha/src/torii.rs:138]: I would suggest to move it into Torii configuration.

File: iroha_data_model/src/lib.rs [hunk 2]
```diff
@@ -1073,6 +1087,9 @@ pub mod transaction {
     use parity_scale_codec::{Decode, Encode};
     use std::{iter::FromIterator, time::SystemTime, vec::IntoIter as VecIter};
 
+    /// Maximum number of instructions and expressions per transaction
+    pub const MAX_INST_NUMBER: usize = 4096;
```
Comment (eadventurous) [inline iroha_data_model/src/lib.rs:1091]: `MAX_INSTRUCTION_NUMBER` and I also suggest to move it into Torii config.

File: iroha/src/tx.rs [hunk 3]
```diff
@@ -12,11 +12,13 @@ use std::{
...
 };
 
-//TODO: check the number of ISI in tx and drop if it exceeds the limit. Also drop if it exceeds the byte limit.
+//TODO: drop transaction it exceeds the byte limit.
 impl TryFrom<Transaction> for AcceptedTransaction {
     type Error = String;
 
     fn try_from(t: Transaction) -> Result<Self, Self::Error> {
+        t.check_instruction_len()
+            .map_err(|e| format!("Failed to accept transaction: {}", e))?;
```
Comment (eadventurous) [inline iroha/src/tx.rs:21]: I think you should add a test for the `TryFrom`.

**Review by eadventurous — CHANGES_REQUESTED** [review]

File: iroha/src/genesis.rs [hunk 4]
```diff
@@ -47,21 +47,23 @@ impl GenesisTransaction {
...
     ) -> Result<AcceptedTransaction, String> {
         let transaction = Transaction::new(
             self.isi.clone(),
             <Account as Identifiable>::Id::genesis_account(),
             GENESIS_TRANSACTIONS_TTL_MS,
         )
-        .sign(&genesis_key_pair)?;
-        AcceptedTransaction::try_from(transaction)
+        .sign(&genesis_key_pair, max_instruction_number)?;
+        AcceptedTransaction::accept(transaction, max_instruction_number)
```
Comment (eadventurous) [inline iroha/src/genesis.rs:58]: I think `transaction.accept(...)` would be better

File: iroha_data_model/src/lib.rs [hunk 5]
```diff
@@ -1126,10 +1143,30 @@ pub mod transaction {
...
         /// Sign transaction with the provided key pair.
         ///
         /// Returns `Ok(Transaction)` if succeeded and `Err(String)` if failed.
-        pub fn sign(self, key_pair: &KeyPair) -> Result<Transaction, String> {
+        pub fn sign(
+            self,
+            key_pair: &KeyPair,
+            max_instruction_number: usize,
+        ) -> Result<Transaction, String> {
+            self.check_instruction_len(max_instruction_number)?;
```
Comment (eadventurous) [inline iroha_data_model/src/lib.rs:1169]: I think there should not be any check on sign. As it is unrelated operation. If you want to add the check to the client library, you can simply check it on submit.

**Review by eadventurous — CHANGES_REQUESTED** [review]

File: iroha/src/genesis.rs (hunk 4 above)
Comment (i1i1) [reply iroha/src/genesis.rs:58]: `AcceptTransaction` is in `iroha` crate while `Transaction` is in `iroha_data_model`.

File: iroha_data_model/src/lib.rs (hunk 5 above)
Comment (i1i1) [reply iroha_data_model/src/lib.rs:1169]: Ok, will remove this check here

File: iroha/src/genesis.rs (hunk 4 above)
Comment (i1i1) [reply iroha/src/genesis.rs:58]: Guess it won't do as AcceptedTransaction type is in iroha crate while Transaction is in iroha_data_model crate. Should I move AcceptedTransaction to iroha_data_model?

Comment (i1i1) [reply iroha/src/genesis.rs:58]: Should we again introduce Accept trait in order to implement this behaviour?

Comment (eadventurous) [reply iroha/src/genesis.rs:58]: Can we simply implement Transaction with `.accept` method in iroha_data_model crate?

Comment (i1i1) [reply iroha/src/genesis.rs:58]: No, we can't reference `iroha`'s crate types inside `iroha_data_model` crate. That is why before there was `Accept` trait in `iroha` crate. It helped to write method for external type not owned by crate.

Comment (eadventurous) [reply iroha/src/genesis.rs:58]: Ok, I see, the accept trait is a bit of an overkill to my mind here. Then I would suggest to rename the method into `AcceptedTransaction::from_transaction()`

File: iroha_client/src/client.rs [hunk 6]
```diff
@@ -65,6 +67,7 @@ impl Client {
     /// Returns submitted transaction's hash or error string.
     pub fn submit_all(&mut self, instructions: Vec<Instruction>) -> Result<Hash, String> {
         let transaction = self.build_transaction(instructions)?;
+        transaction.check_instruction_len(self.max_instruction_number)?;
         self.submit_transaction(transaction)
```
Comment (eadventurous) [inline iroha_client/src/client.rs:71]: It's better to check in `submit_transaction` method to my mind.

**Review by eadventurous — APPROVED** [review]

**Pull request merged**